neqo-http3/src/push_controller.rs

```diff
@@ -1,20 +1,91 @@
+        headers: Vec<u8>,
+    },
+}
+
+// PushController keeps information about push stream states.
+//
+// PushStates:
+//   Init: there is no push stream nor a push_promise. This state is only used to keep track of opened and closed
+//         push streams.
+//   PushPromise: the push has only ever receive a push_promise frame
+//   OnlyPushStream: there is only a push stream. All push stream events, i.e. PushHeaderReady and PushDataReadable
+//                   will be delayed until a push_promise is received (they are kept in `events`).
+//   Active: there is a push steam and at least one push_promise frame.
+//
+// push_ids smaller than next_push_id_to_open are all in one of the above state or they are closed. The closed pushes
+// are removed from push_streams. push_ids >= next_push_id_to_open have not been opened yet.
+//
+// A PushStream calls `add_new_push_stream` that may change the push state from Init to OnlyPushStream or from
+// PushPromise to Active. If a stream has already been closed `add_new_push_stream` returns false (the PushStream
+// will close the transport stream).
+// A PushStream calls `push_stream_reset` if the transport stream has been canceled.
+// When a push stream is done it calls `close`.
+//
+// The PushController handles:
+//  PUSH_PROMISE frame: frames may change the push state from Init to PushPromise and from OnlyPushStream to
+//                      Active. Frames for a closed steam is ignored.
+//  CANCEL_PUSH frame: (`handle_cancel_push` will be called). If a push is in state PushPromise or Active, any
+//                     posted events will be removed and a push_canceled event will be posted. If a push is in
+//                     state `OnlyPushStream` or `Active` the transport stream and the `PushStream` will be closed.
+//                     The frame will be ignored for already closed pushes.
+//  Application calling cancel: the actions are similar to the CANCEL_PUSH frame. The difference is that
+//                              `push_canceled` will not be posted and a CANCEL_PUSH frame may be sent.
 
 #[derive(Debug)]
-pub(crate) struct PushController {}
+pub(crate) struct PushController {
+    max_concurent_push: u64,
+    current_max_push_id: u64,
+    // push_streams holds the states of push streams.
+    // We keep a stream until the stream has been closed.
+    push_streams: BTreeMap<u64, PushState>,
+    // The keeps the next consecutive push_id that should be open.
+    // All push_id < next_push_id_to_open are in the push_stream lists. If they are not in the list they have
+    // been already closed.
+    next_push_id_to_open: u64,
+    conn_events: Http3ClientEvents,
+}
 
 impl PushController {
-    pub fn new() -> Self {
-        PushController {}
+    pub fn new(max_concurent_push: u64, conn_events: Http3ClientEvents) -> Self {
+        PushController {
+            max_concurent_push,
+            current_max_push_id: 0,
+            push_streams: BTreeMap::new(),
```

Comment:
I think that this would be better as a `VecDeque`.  It requires keeping one extra piece of state (the push_id of the first item), but it would be far more performant and some of the other utilities would be nicer.  You could wrap it in a small struct so that indexing is easy.

Reply:
For this I will need a Close state as well. It is doable.